docs: ABFT non-normative soft vote (#138)
algorandfoundation/specs: src/abft-nn-soft-vote.md

\newcommand \RetrieveProposal {\mathrm{RetrieveProposal}}
\newcommand \DynamicFilterTimeout {\mathrm{DynamicFilterTimeout}}
\newcommand \Soft {\mathit{soft}}
\newcommand \Vote {\mathrm{Vote}}
\newcommand \function {\textbf{function }}
\newcommand \endfunction {\textbf{end function}}
\newcommand \if {\textbf{if }}
\newcommand \then {\textbf{ then}}
\newcommand \endif {\textbf{end if}}
\newcommand \for {\textbf{for }}
\newcommand \do {\textbf{ do}}
\newcommand \endfor {\textbf{end for}}
\newcommand \loh {\mathit{lowestObservedHash}}
\newcommand \vt {\mathit{vote}}
\newcommand \ph {\mathit{priorityHash}}
\newcommand \c {\mathit{credentials}}
\newcommand \prop {\mathit{proposal}}
$$

# Soft Vote

The soft vote stage (also known as _"filtering"_) filters the proposal-value candidates 
available for the round, selecting the one with the highest priority to vote for.

## Priority Function

Let \\( \Priority \\) be the function that determines which proposal-value to
_soft-vote_ for this round, as defined in the [normative section](./abft-player-state.md#special-values):

<!-- TODO: Replace with anchored include once the normative section is merged -->
$$

> \Priority(v) = \min_{i \in [0, w_j)} \left\\{ \Hash \left( \VRF.\ProofToHash(y) || i \right) \right\\}

Reviewer: Just realized we missed the hashed address here in the Priority function definition,
```suggestion
\Priority(v) = \min_{i \in [0, w_j)} \left\\{ \Hash \left( \VRF.\ProofToHash(y) || I_j || i \right) \right\\}
```

$$

Where:

> - \\( w_j \\) be the weight of the credentials for \\( v \\) by player \\( I_j \\),

Reviewer: Adding explicit reference to the address here as it is now included in the function
```suggestion
-\\( I_j \\) be the address of a distinct player identified by the subscript \\( j \\).
- \\( w_j \\) be the weight of the credentials for \\( v \\) by player \\( I_j \\),
```

Reviewer: Wrt language: since we are not using "let", should we just say e.g. "\\( w_j \\) *is* the weight of..." instead of "\\( w_j \\) *be* the weight of..."?

> - \\( y \\) be the result of the signing procedure for \\( v \\) by player \\( I_j \\).

Reviewer: Per discussed with @cusma earlier today:
```suggestion
- \\( y \\) be the VRF proof as computed by player \\( I_j \\) using their VRF secret key.
```

The function selects the minimum among a set of \\( w_j \\) hash values calculated

> as \\(\Hash \left(\VRF.\ProofToHash(y) || i \right)\\), with \\(i \in [0, w_j)\\).

Reviewer: ```suggestion
as \\(\Hash \left(\VRF.\ProofToHash(y) || I_j || i \right)\\), with \\(i \in [0, w_j)\\).
```

The higher the credentials’ weight \\( w_j \\), the larger the set, the higher the
chances for the player \\( I_j \\) to get the lowest value among all the players
for this round.

## Algorithm

---

\\( \textbf{Algorithm 4} \text{: Soft Vote} \\)

$$
\begin{aligned}
&\text{1: } \function \SoftVote() \\\\
&\text{2: } \quad \loh \gets \infty \\\\
&\text{3: } \quad v \gets \bot \\\\
&\text{4: } \quad \for \vt^\prime \in V \text{ with } \vt^\prime_s = \prop \do \\\\

> &\text{5: } \quad \quad \ph \gets \Priority(\vt^\prime)  \\\\

Reviewer: ```suggestion
&\text{5: } \quad \quad \ph \gets \Priority(\vt^\prime_v)  \\\\
```

&\text{13:} \quad \quad \quad \c \gets \Sortition(a_I, r, p, \Soft) \\\\
&\text{14:} \quad \quad \quad \if \c_j > 0 \then \\\\
&\text{15:} \quad \quad \quad \quad \Broadcast(\Vote(a_I, r, p, \Soft, v, \c)) \\\\
&\text{16:} \quad \quad \quad \quad \if \RetrieveProposal(v) \then \\\\
&\text{17:} \quad \quad \quad \quad \quad \\Broadcast(\RetrieveProposal(v)) \\\\
&\text{18:} \quad \quad \quad \quad \endif \\\\
&\text{19:} \quad \quad \quad \endif \\\\
&\text{20:} \quad \quad \endfor \\\\
&\text{21:} \quad \endif \\\\
&\text{22: } \endfunction
\end{aligned}
$$

---

> ⚙️ **IMPLEMENTATION**
>
> Soft vote filtering [reference implementation](https://github.com/algoradam/go-algorand/blob/master/data/committee/credential.go#L160C1-L187C2).
> Soft vote issuance [reference implementation](https://github.com/algorand/go-algorand/blob/df0613a04432494d0f437433dd1efd02481db838/agreement/player.go#L170-L206).

The soft vote stage is run after a timeout of \\( \DynamicFilterTimeout(p) \\)
(where \\( p \\) is the executing period of the node) is observed by the node (see
the [dynamic filter timeout section](./abft-nn-dynamic-filter-timeout.md) for more
details).

> Let \\( V^\ast \\) be all proposal votes received, e.g., \\( V^\ast = \\{vt^\prime \in V : vt^\prime_s = \prop\\} \\).

Reviewer: As per discussed with @cusma, suggestion to move the paragraph to right over the Algorithm (at the end of the previous section). Adding a bit of context and slightly rephrasing to fit in its new spot:
```suggestion
Let \\( V \\) be the set of all observed votes in the currently executing round. For convenience, we define a subset, \\( V^\ast \\) to be all proposal votes received; that is \\( V^\ast = \\{vt^\prime \in V : vt^\prime_s = \prop\\} \\).
```

With the aid of a priority hash function (see the [normative section](./abft.md#special-values)),
this stage performs a filtering action, keeping the lowest hashed value observed.

The priority function (**Algorithm 4** - Lines 4 to 9) should be interpreted as
follows.

> Consider every proposal \\( \vt \\) in \\( V^\ast \\). Given the sortition

Reviewer: ```suggestion
Consider every proposal value \\( \v \\) in votes belonging to \\( V^\ast \\). Given the sortition
```
